File: csharp/tests/VerdictRules.Tests/TestHelpers.cs
namespace VerdictRules.Tests;

/// <summary>
/// Test-rule builders shared across the suite. <see cref="Pass"/>/<see cref="Fail"/>
/// mirror Python's own local `_pass`/`_fail` helpers in `test_rule.py` and
/// `test_engine.py`; <see cref="Counting"/> is the call-log rule every
/// short-circuit proof in this suite (and in `CSharpIdiomTests.cs`) is built on.
/// </summary>
internal static class Rules
{
    /// <summary>A rule that always passes.</summary>
    public static FunctionRule Pass(string name, string? group = null, object? data = null) =>
        new(name, (_, _) => Task.FromResult(new RuleResult(name, true, data: data)), group);

    /// <summary>A rule that always fails.</summary>
    public static FunctionRule Fail(string name, string? group = null, string detail = "") =>
        new(name, (_, _) => Task.FromResult(new RuleResult(name, false, detail)), group);

    /// <summary>
    /// A rule that records every evaluation, so short-circuiting can be proven
    /// by what actually ran rather than by the final boolean alone — a port
    /// that evaluated concurrently would return the same boolean and fail
    /// only here.
    /// </summary>
    public static FunctionRule Counting(string name, bool passes, List<string> log, string? group = null) =>
        new(name, (_, _) =>
        {
            log.Add(name);
            return Task.FromResult(new RuleResult(name, passes));
        }, group);

    /// <summary>An empty context — most tests here don't care what's in it.</summary>
    public static readonly IReadOnlyDictionary<string, object?> Empty =
        new Dictionary<string, object?>();
}

File: dart/packages/verdict_rules/test/test_helpers.dart
import 'package:verdict_rules/verdict_rules.dart';

/// A rule that always passes, mirroring Python's own `_pass` test helper.
FunctionRule pass(String name, {String? group, Object? data}) => FunctionRule(
      name,
      (ctx) async => RuleResult(ruleName: name, passed: true, data: data),
      group: group,
    );

/// A rule that always fails, mirroring Python's own `_fail` test helper.
FunctionRule failing(String name, {String? group, String detail = ''}) =>
    FunctionRule(
      name,
      (ctx) async => RuleResult(ruleName: name, passed: false, detail: detail),
      group: group,
    );

/// A rule that records every evaluation, so short-circuiting can be proven by
/// what actually ran rather than by the final boolean alone. A port that
/// evaluated concurrently would return the same boolean and fail only here.
FunctionRule counting(
  String name,
  bool passes,
  List<String> log, {
  String? group,
}) =>
    FunctionRule(name, (ctx) async {
      log.add(name);
      return RuleResult(ruleName: name, passed: passes);
    }, group: group);
